src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -587,25 +619,34 @@ internal static string GetServerDefinedTypeFullName(Type type)
         /// <returns>Client clr name.</returns>
         internal static string GetClientFieldName(Type t, string serverDefinedName)
         {
-            List<string> serverDefinedNames = serverDefinedName.Split(',').Select(name => name.Trim()).ToList();
-            List<string> clientMemberNames = new List<string>();
-            foreach (var serverSideName in serverDefinedNames)
+            ODataTypeInfo typeInfo = GetODataTypeInfo(t);
+            string clientFieldName;
+
+            if (!typeInfo.ClientDefinedNameDict.TryGetValue(serverDefinedName, out clientFieldName))

```

Comment:
>ClientDefinedNameDict [](http://example.com/codeflow?start=26&length=21)

note that serverDefinedName is a comma-separated list of property names.  So your cache is caching that set of field names, not the indidual fields, and if the code in one point asks for type a, in another point asks for type b, in another point asks for a, b, and in another point asks for b,a, those are all distinct computed entries in your cache where you really should just have 2.